### PR TITLE
Improve recovery on malformed format call

### PR DESCRIPTION
**Pull request by scrabsha**
The token following a format expression should be a comma. However, when it is replaced with a similar token (such as a dot), then the corresponding error is emitted, but the token is treated as a comma, and the parsing step continues.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/format.rs [hunk 1]
```diff
@@ -165,7 +165,16 @@ fn parse_args<'a>(
                 let mut err = ecx.struct_span_err(p.token.span, "expected token: `,`");
                 err.span_label(p.token.span, "expected `,`");
                 p.maybe_annotate_with_ascription(&mut err, false);
-                return Err(err);
+                match token::TokenKind::Comma.similar_tokens() {
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/format.rs:168]: This recovery is currently under the `if first` condition, but it is applicable to all commas in the list, not only the first one.

Comment (scileo) [reply compiler/rustc_builtin_macros/src/format.rs:168]: Nice catch! I'll send a patch soon.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 2]
```diff
@@ -160,15 +160,26 @@ fn parse_args<'a>(
...
                 // After `format!(""` we always expect *only* a comma...
                 let mut err = ecx.struct_span_err(p.token.span, "expected token: `,`");
                 err.span_label(p.token.span, "expected `,`");
                 p.maybe_annotate_with_ascription(&mut err, false);
-                return Err(err);
+                err
             } else {
                 // ...after that delegate to `expect` to also include the other expected tokens.
-                let _ = p.expect(&token::Comma)?;
+                p.expect(&token::Comma).unwrap_err()
```
Comment (scileo) [inline compiler/rustc_builtin_macros/src/format.rs:171]: I am unsure that this call of `unwrap_err` won't panic. Is there a better option?

Comment (petrochenkov) [reply compiler/rustc_builtin_macros/src/format.rs:171]: A better option is to always use `p.expect()`, but do
```rust
if first {
    p.expected_tokens.clear();
}
```
before that.

`expected_tokens` is private though, but you can add a public `clear_expected_tokens` method to `Parser`.

Comment (scileo) [reply compiler/rustc_builtin_macros/src/format.rs:171]: Nice, thanks for the tip.

**Comment by petrochenkov** [comment]
Looks like one of UI tests wasn't updated.

**Comment by scileo** [comment]
> Looks like one of UI tests wasn't updated.

This has been fixed in my last force-push.

**Pull request merged**